0.5.x

jost125 opened the pull request.

jost125 committed:
[BC] Rework applicatives, monads, function currying

This introduces some small abstraction over applicatives and mondas to
help us build new typeclasses faster.

Also I have reworked function currying to be able to do typehinting much
better for curried functions

jost125 committed:
Configure slevomat rule to be PSR12 compliant

jost125 committed:
Remove unhelping rule

It is already checked with phpstan (and better) and it also didn't allow
ommiting var name (which is totally valid)

jost125 committed:
Upgrade phpstan

jost125 committed:
Add some doc on monad/applicative abstraction

jost125 committed:
[BC] Bump php version to 7.4

jost125 committed:
Disable check for arrow function

Becuase we are on 7.4 as min version, yay

jost125 committed:
Disable 120 chars for comments

To fit long php doc types, namely for thouse complex phpstan types

jost125 committed:
Support lifting with up to 30 fixed arguments

This allows quite good type safety (it check passed callable for number
of arguments and creates well typed callable in return)

jost125 committed:
Support currying with up to 30 fixed arguments

honzatrtik commented on src/Bonami/Collection/CurriedFunction.php:
```diff
@@ -61,34 +61,1010 @@ public static function of(callable $callable): self
+     * @param callable(I1, I2, I3, I4, I5, I6, I7, I8, I9, I10, I11, I12, I13, I14, I15, I16, I17, I18, I19, I20, I21, I22, I23, I24, I25, I26, I27, I28, I29, I30): Z $callable
+     *
+     * @return self<I1, self<I2, self<I3, self<I4, self<I5, self<I6, self<I7, self<I8, self<I9, self<I10, self<I11, self<I12, self<I13, self<I14, self<I15, self<I16, self<I17, self<I18, self<I19, self<I20, self<I21, self<I22, self<I23, self<I24, self<I25, self<I26, self<I27, self<I28, self<I29, self<I30, Z>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
+     */
```
Wow, 30! thats a lot ;-)

Do you think there is a use case for parameter list such long? Maybe some autogenerated code?

jost125 replied:
Scala has limit of 23 for similar purposes. I have sometime hit that limit in some circumstances. There might be use cases for some auto generated code, or for some factory methods of some very flat structures, that might want very long lists of arguments.

honzatrtik commented on src/Bonami/Collection/Applicative1.php:
```diff
@@ -0,0 +1,106 @@
+     *
+     * @return self<B>
+     */
+    abstract public static function ap(self $closure, self $argument): self;
```
The more I think about it, more I am convinced that (especially for newcomers) its better to define Applicative in terms of `product` function (or some variant of it) -
in ts it'd be roughly like this (for `Option`)
```ts
<A,B>(fa: Option<A>, fb: Option<B>) => Option<[A, B]>
```

see https://typelevel.org/cats/typeclasses/applicative.html#applicative for details

Also it would not need to be static.. 
```php
    /**
     * @template B
     *
     * @param self<B> $fb
     * @return self<array{0: T, 1: B}>
     */
abstract public function butNotSureAboutTheName(self $fb): self;
```

On the other hand, not sure, how comfortable the work with tuples is in php at the moment..

honzatrtik reviewed (approved):
👍 gj 🎉

jost125 committed:
Remove phpstan prefixes

They are no longer needed

jost125 committed:
Update doc for function currying

jost125 committed:
Use shorter php 8.1 syntax in docs

jost125 committed:
Add previous exception to failure

This can be handy for debugging

jost125 committed:
Implement Iterable1 typeclass

This way we can easily prototype Iterable structures and ensure same
interface on multiple stuctures without actual interface.

Concreate class using the trait should override methods for optimization
when needed.

jost125 committed:
Use Iterable1 in our structs

jost125 committed:
Document Iterable1

jost125 committed:
Separate and update Option doc into own file

jost125 committed:
Separate and update TrySafe doc

jost125 committed:
Separate lift into own document

jost125 committed:
Add alternative applicative definition via product

jost125 committed:
Derive ap implementation for product in Applicative

jost125 committed:
Implement traverse via product

It is more obvious and more performant then via `ap`.

jost125 committed:
Implement method for immutable item addition

jost125 committed:
Drop somehow forgotten passage about type classes

jost125 committed:
Update doc a bit

jost125 committed:
Allow some composer plugins

jost125 committed:
Implement doWhile for lazy lists
